### ISSUE TITLE
Rule proposal: require at least two arguments for `setTimeout()` and `setInterval()`

### ISSUE THREAD

**Issue by Trott** [eslint/eslint#7542]
**Please describe what the rule should do:**

`setTimeout()` and `setInterval()` can be used with just a callback. In Node.js, the omitted duration (second argument) will use a default value of 1ms. In browsers, it does the same thing or something very similar.

However, omitting the duration makes intention unclear. Was it omitted by accident? Did the author intend to use `setImmediate()` instead? And so on.

So this rule would flag uses of `setTimeout()` and `setImmediate()` if there are not at least two arguments provided (the callback and the duration).

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
setTimeout(function () {
  console.log('foo');
  bar += baz;
});

setInterval(callback);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a widely-applicable, generic, and environment-agnostic rule.

**Comment by Trott** [comment]
I should add that I have implemented this rule and proposed it as a custom rule for Node.js core but would much prefer it be a built-in rule. https://github.com/nodejs/node/pull/9472/commits/04043b190ccdfcc9a442889879ce50155b6f698b

**Comment by mysticatea** [comment]
Thank you for this issue.

Sounds fine to me since we have similar rule [radix](http://eslint.org/docs/rules/radix) in core.

**Comment by platinumazure** [comment]
Personally I'm on the fence about whether ESLint is incomplete without this rule (our standard for inclusion of new rules since 2016). I won't stand in the way if the rest of the team agrees this is worth implementing, but if not, this would make an excellent foundation to a timeout-based eslint plugin.

**Comment by nzakas** [comment]
I'm more of the mind that this should be a plugin rule. `setTimeout` and `setInterval` are not parts of ECMAScript (unlike `parseInt()`), and I don't think we want to get into the business of checking arguments for all common methods.

**Comment by mysticatea** [comment]
Ah, true. I had forgotten that ECMAScript does not include `setTimeout` and `setInterval`.

**Comment by kaicataldo** [comment]
I agree that this sounds useful but should be a plugin rule, for the same reasons listed above.

**Comment by ilyavolodin** [comment]
@Trott thanks for the suggestion. It looks like the team feels pretty strongly that this should be done outside of the core as a plugin. Feel free to reach out to us in our [Gitter](https://gitter.im/eslint/eslint) chatroom if you need help creating one. I'm closing this issue, as with so many people on the team opposing this change, it will not make it into the core.
